mesonbuild/interpreter.py

```diff
@@ -2263,6 +2264,42 @@ def _get_native() -> str:
             else:
                 return _get_native()
 
+    @permittedKwargs({'encoding', 'allow_absolute'})
+    @FeatureNew('meson.get_file_contents', '0.58.0')
+    def get_file_contents_method(
+            self, args: T.Sequence[str], kwargs: dict) -> str:
+        '''
+        Read a file from the source tree and return its value as a decoded
+        string. If the encoding is not specified, the file is assumed to be
+        utf-8 encoded. Paths must be relative by default (to prevent accidents)
+        and are forbidden to be read from the build directory (to prevent build
+        loops)
+        '''
+        if len(args) != 1:
+            raise InterpreterException('expected single positional <path> arg')
+        path = args[0]
+        encoding = kwargs.get('encoding', 'utf-8')
+        allow_absolute = kwargs.get('allow_absolute', False)
+        source_dir = self.interpreter.environment.get_source_dir()
+        build_dir = self.interpreter.environment.get_build_dir()
+        if not allow_absolute and os.path.isabs(path):
+            raise InterpreterException('paths must be relative')
+        path = os.path.join(source_dir, path)
+        if mesonlib.path_is_in_root(Path(path), Path(build_dir), resolve=True):
+            raise InterpreterException('path must not be in the build tree')
+        try:
+            with open(path, 'rb') as f:
+                data = f.read().decode(encoding)
```

Comment:
you can save yourself an extra decode by passing the encoding to `open()`:

```python
with open(path, 'r', encoding=encoding) as f:
     data = f.read()`
```
And you really don't need the rest of this in the context manager or in the try block.

Reply:
This is deliberate. I keep the rest in the try block because I don't want to create a dependency on the file if we fail to decode it

Reply:
But you raise an uncaught exception if the decode fails, so no dependency will be added. Unless I'm missing something.